### PR TITLE
Add per-image GT vs pred comparison image during validation

### PR DESCRIPTION
**Pull request by ShuaiLYU**
This pull request adds functionality to save side-by-side visualizations of ground truth and predicted bounding boxes for each image during YOLO model validation. This enhancement helps with qualitative analysis by making it easier to compare model predictions with actual labels. The main changes are as follows:

**Visualization Enhancements:**

* Added a new private method `_save_per_image_gt_pred` to the `val.py` validation logic, which creates and saves side-by-side images showing both ground truth and prediction panels for each sample. This uses the same drawing logic as `plot_images` and saves the result in a dedicated directory.
* Updated the `update_metrics` method to call `_save_per_image_gt_pred` for each sample when not in training mode and on the main process, enabling per-image visualization during validation runs.

**Dependency Update:**

* Imported `Image` from the `PIL` library to support image saving functionality.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds automatic per-image ground-truth vs prediction comparison images during detection validation for easier visual inspection of model results. 🖼️

### 📊 Key Changes
- Introduces a new `_save_per_image_gt_pred()` helper in `ultralytics/models/yolo/detect/val.py`.
- Saves side-by-side GT and Pred visualizations for each validation image into `save_dir/gt_pred_per_image`.
- Reuses existing `plot_images()` rendering logic for consistency between standard validation plots and the new per-image outputs.
- Triggers image export during validation only when not training and only on the main process (`RANK in {-1, 0}`) to avoid duplicate saves in distributed runs.
- Adds `PIL.Image` usage to concatenate and save the final comparison image.

### 🎯 Purpose & Impact
- Makes it much easier to inspect prediction quality on a per-image basis without manually comparing separate outputs. 🔍
- Helps users debug false positives, false negatives, and localization issues more efficiently.
- Improves validation interpretability for detection workflows with minimal impact on existing behavior.
- Useful for benchmarking, qualitative review, and dataset/model error analysis across YOLO detection tasks.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ShuaiLYU, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness, and an engineer will assist you shortly. Your per-image GT|Pred validation visualization update looks very helpful for qualitative inspection 🖼️

To ensure a seamless integration of your work, please review the following checklist:
- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/models/yolo/detect/val.py [hunk 1]
```diff
@@ -193,6 +194,8 @@ def update_metrics(self, preds: list[dict[str, torch.Tensor]], batch: dict[str,
                 self.confusion_matrix.process_batch(predn, pbatch, conf=self.args.conf)
                 if self.args.visualize:
                     self.confusion_matrix.plot_matches(batch["img"][si], pbatch["im_file"], self.save_dir)
+                if not self.training and RANK in {-1, 0}:
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/detect/val.py:197]: ⚠️ **HIGH**: This runs for every image in standalone validation regardless of `plots=False`, so a normal `model.val()` will now do two extra renders plus an image write per sample. On larger validation sets that is a substantial regression in runtime and disk usage, and it also bypasses the existing user-facing switch for disabling plots. Please gate this behind `self.args.plots` (or a dedicated opt-in arg) so the extra work is disableable.

**Suggested change:**
```suggestion
                if self.args.plots and not self.training and RANK in {-1, 0}:
```

File: ultralytics/models/yolo/detect/val.py [hunk 2]
```diff
@@ -336,6 +339,58 @@ def plot_val_samples(self, batch: dict[str, Any], ni: int) -> None:
...
+            paths=[pbatch["im_file"]],
+            names=self.names,
+            save=False,
+            conf_thres=self.args.conf,
+            threaded=False,
+        )
+        if gt_img is None or pred_img is None:
+            return
+
+        Image.fromarray(np.concatenate((gt_img, pred_img), axis=1)).save(save_dir / f"{image_path.stem}_gt_pred.jpg")
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/detect/val.py:404]: 💡 **MEDIUM**: Using only `image_path.stem` will silently overwrite outputs when the validation set contains duplicate basenames in different directories, which is common in nested datasets. The saved comparison should use a path-derived unique name instead of just the stem.

**Suggested change:**
```suggestion
        Image.fromarray(np.concatenate((gt_img, pred_img), axis=1)).save(save_dir / f"{image_path.as_posix().replace('/', '_').replace(':', '_').rsplit('.', 1)[0]}_gt_pred.jpg")
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Useful feature and the helper cleanly reuses `plot_images`, but there are two issues to address before merge: the export path is not uniquely named, and the new per-image rendering is always enabled for standalone validation, creating an avoidable runtime/disk regression.

💬 Posted 2 inline comments

**Comment by ShuaiLYU** [comment]
@Laughing-q 
I’ve added a feature to save GT-prediction comparison images per image during validation.
This is a helpful utility for identifying bad cases and debugging model predictions more intuitively.

**Comment by ShuaiLYU** [comment]
Okay, just fixed a compatibility issue with the OBB task.
```
yolo obb val model=yolo26n-obb.pt data=dota8.yaml  plots=True     
```
here, I show one of the images in [save_dir]/gt_pred_per_image 

image

**Comment by glenn-jocher** [comment]
Hi @ShuaiLYU — I think this is already covered in `main`. `ConfusionMatrix.plot_matches` (`ultralytics/utils/metrics.py:497`) renders the per-image GT / TP / FP / FN panels, and it's wired into detect validation via `visualize=True` at `ultralytics/models/yolo/detect/val.py:197`. That landed in merged PR #18868 (commit `b905114f46`), which predates this PR, so as-is this duplicates existing functionality.

If there's a concrete capability your version adds that `plot_matches` doesn't, could you describe it and rebase to **extend `plot_matches`** rather than add a parallel path? If not, let's close this one. Thanks!

**Comment by glenn-jocher** [comment]
Thank you @ShuaiLYU for this contribution! Per-image GT/TP/FP/FN comparison visualization shipped on main in #18868 (`ConfusionMatrix.plot_matches`, wired into validation via `plots`/`visualize` args) — see https://docs.ultralytics.com/modes/val/. Since the branch now carries no remaining changes, I'm closing this one. We appreciate the idea and effort, and would love to see future contributions!
